src/mpid/ch4/generic/am/mpidig_am_send_utils.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIDIG_AM_SEND_UTILS_H_INCLUDED
+#define MPIDIG_AM_SEND_UTILS_H_INCLUDED
+
+/* This file is for supporting routines used for pipelined data send. These routines mainly is for
+ * managing the send request counters, completion counters and DT refcount */
+
+/* prepare the send counters for size of data and datatype that we need to refcount */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_send_cntr_init(MPIR_Request * sreq, MPI_Datatype datatype,
+                                                    size_t data_sz)
+{
+    MPIDIG_sreq_send_cntr_t *cntr = &MPIDIG_REQUEST(sreq, req->send_cntr);
```

Comment:
This essentially is the send-side counter part of the "async" struct on the recv-side. Maybe we should put both into a union and name them in a more consistent manner.

Reply:
We could do that. It is already in a union, just the name is called `send_cntr`. Maybe we can call it `send_async` and do a rename for the existing async in a different PR.

In this way, we can probably rename the `MPIDIG_send_cntr_` functions to just `MPIDIG_send_` functions?